### ISSUE TITLE
Running into nans or encountering segmentation faults during training

### ISSUE THREAD

**Issue by ashwinipokle** [kuleshov-group/mdlm#7]
Hello,

Thanks for this great work. I'm training this model on some simple datasets and while training, after 2-3 epochs, the model either outputs nans (likely due to mixed precision training, happens frequently with D3PM and SEDD) or with SUBS parametrization, I run into segmentation fault in these lines in `diffusion.py`.

    unmasked_indices = (xt != self.mask_index)
    logits[unmasked_indices] = self.neg_infinity
    logits[unmasked_indices, xt[unmasked_indices]] = 0 

Do you have any suggestions for training strategies to get around / avoid these issues? 

Thanks!

**Comment by s-sahoo** [comment]
I believe you’re using `float16` precision which is causing the underflow. In that case, you should lower the magnitude of `self.neg_infinity` to something like `-1000`.

However, please note that you must use bfloat16 to reproduce the numbers in the paper.

**Comment by ashwinipokle** [comment]
In the config, I have set   `precision: 'bf16'`. Besides that, I'm using the default values originally in the code for other `torch.cuda.amp.autocast` such as here in DiT forward pass https://github.com/kuleshov-group/mdlm/blob/master/models/dit.py#L365. One autocast that might be problematic is this one which autocasts things to float32: https://github.com/kuleshov-group/mdlm/blob/master/diffusion.py#L315. Do you recommend explicitly setting dtype of DiT to bfloat16?

**Comment by s-sahoo** [comment]
I do recommend that you keep the autocasts as they are. What GPU are you using?

**Comment by ashwinipokle** [comment]
I'm using Nvidia A6000 with CUDA 12.4.

**Comment by s-sahoo** [comment]
We verified that our code runs without any NaNs on `A5000`, `A100`, and `H100` GPUs. I’m not sure if we can be of much help here, I’m sorry. One workaround to address NaNs is to `set self.neg_infinity = -1000`, but this might result in slightly off (~3%) perplexity numbers.

**Comment by ashwinipokle** [comment]
Sounds good! Let me try that and check if it fixes training.
